### PR TITLE
Some changes to `Provider` and `Property` for more convenient use from Kotlin and Groovy

### PR DESCRIPTION
**Pull request by adammurdoch**
tl;dr

```kotlin
someTask.someProp.set("abc")
// becomes
someTask.someProp.value = "abc"

consumer.someInput.set(producer.someOutput)
// becomes
consume.someInput.source = producer.someOutput

someProp.get().asFile.writeText(otherProp.get())
// becomes
someProp.value.asFile.writeText(otherProp.value)
```

This is a proposal for some changes to to `Provider` and `Property`. Specifically:

- Add read-only `value` property to `Provider<T>` to represent the current value of the provider. This is the same as `get()`.
- Add mutable `value` property to `Property<T>` to allow the current value of the property to be set and queried. This is the same as `get()` and `set(T)`.
- Add mutable `source` property to `Property<T>` to allow the current _source_ of the property to be set. This is the same as `this` and `set(Provider<T>)`.

The intention is that `get()` and `set()` would be deprecated and replaced by these new methods.
None of these properties are nullable. Instead, a method to explicitly discard the value would be added (not included in this change yet).

The intention is also that the Groovy DSL convenience would be deprecated.

Similar changes would be made to the other property types. For example:

```kotlin
someTask.input.fileProvider(legacyTask.map { it.outputFile})
// becomes
someTask.input.fileSource = legacyTask.map { it.outputFile }
```

### Contributor Checklist
- [ ] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [ ] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#git-commit---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [ ] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/core-api/src/main/java/org/gradle/api/provider/Property.java [hunk 1]
```diff
@@ -45,6 +45,24 @@
...
+     * @since 6.1
+     */
+    @Incubating
+    void setValue(T value);
+
+    /**
+     * @since 6.1
+     */
+    @Incubating
+    Provider<? extends T> getSource();
```
Comment (adammurdoch) [inline subprojects/core-api/src/main/java/org/gradle/api/provider/Property.java]: This is required to make Kotlin happy. It does not consider `source` a property unless there is a getter method.

**Comment by melix** [comment]
I'm not super convinced: it doesn't make the DSL much better, at least not enough to justify deprecating methods IMO. I'd very much prefer we avoid the many deprecations/API changes we're doing which make the lives of plugin authors painful.

**Comment by eskatos** [comment]
I think it makes the DSL better by removing the parentheses, favoring assignments and better communicating the intent by using actual words instead of the less meaningful `get` and `set`. Distinguishing setting a value or a source of value also makes a lot of sense to me and makes the DSL more explicit, when you read `some.source = ...` you know directly that's it's a lazy wiring, when you read `some.value = ..` you know directly that's it's an eager assignment.

I share Cédric's concerns about the impact of deprecations. Build logic often has a lot of those wiring. But I also think it's worth it because of the reasons above, especially the better communication of intent.

**Comment by melix** [comment]
I'm not saying the proposal doesn't have value. I'm saying I don't see _much_ value compared to breaking existing plugins with deprecations. I'm seeing a lot of swearing on Twitter about breaking changes in APIs we promote, typically the provider API. It's a cost to invest in those new APIs, and changing them is painful as it makes it harder for plugin authors to support a wide range of Gradle versions.

**Comment by lacasseio** [comment]
That may just be me, but I fail to see a clear improvement to what is already present. First, it as much or more typing for the users. Second, it add another layer of confusion regarding how to wire values together. I fear a snowball effect every time a user change from an eager type to a lazy type because of the `value` vs `source`. I get that it's more explicit but for users that don't care about their build system, I think it's just going to add more swearing. I think most improvement such as discarding the value can definitely be achieve through the current API. Finally, I agree that adding more deprecation would just frustrate users even more.

I think going forward with deprecating the "magic" Groovy decorator would be more beneficial then adding new APIs. However, I would argue that we should deprecate the Groovy decorator once we widely use the Provider API in the Gradle codebase to give a progressive migration path to users.

**Comment by big-guy** [comment]
I'll echo some of the other comments.  I think my gut feelings are...
1. We're far down the path of adoption of get/set, so removing those APIs would be very disruptive. 
2. The combination of getters and setters for value and source make the API more confusing to use. I think one of the good things with the existing value(...), convention(...) and set(...) methods is that they can be explained as influencing one "thing" (the eventual value of the provider) where now it looks like there are two different things (a value and a source) plus the eventual value of the provider.
3. Like above, I think it gets worse where we want to provide setters for other types (File -> FileSystemLocation) since it would require more getter/setter pairs to be added. This might be an argument against providing these conveniences on the Provider/Property types.
4. None of the examples are super compelling.  The first two could be done today with value(...):
```
someTask.someProp.set("abc")
// becomes
someTask.someProp.value("abc")

consumer.someInput.set(producer.someOutput)
// becomes
consume.someInput.value(producer.someOutput)
```
I get that this isn't an assignment, but I don't know if assignment is the right way to think about how a Provider gets a value. 

OTOH, I think getValue could be added separately from this as it's slightly better without the complications above.

I think we'd be better off spending time on making it easier to use the existing Provider APIs with older APIs than to migrate to a slightly different Provider API getter/setter syntax.

**Comment by paplorinc** [comment]
Would it make sense to revive and prioritize this PR?

**Comment by YAHSON0** [comment]
> * Check

When I said that the variant was inhuman I meant that this past year has been hell he's been hell I've been feeling like I was held hostage but not physically but spiritually financially my transportation finance everything was struck these people really mess with me bad to where it took a an effect on my personality family doesn't know me anymore I'm getting so angry because they continuously over and over again all night all day all night all day bashing every corner of my account and going for all my repositories all all of the money all the digital wallets I mean until it's drain dry to zero zilch and it was so much money that you couldn't even see it in a lifetime to spend so what I'm saying is it is inhuman to torture and terrorize a human being like that there's violating their constitutional rights is violating their privacy and also destroying them mentally and spiritually you're breaking somebody down specially if the whole group is a bunch of narcissists and out in that so yes they stole everything from me and I want it back they bashed my accounts they they really was very fraudulent evil and very disrespectful without a caring a world of anybody's life

**Comment by YAHSON0** [comment]
Rent a scan on my device see what we can pick up if that's possible what do I have to go
